logstash-core/src/main/java/org/logstash/execution/AbstractPipelineExt.java

```diff
@@ -540,6 +576,23 @@ private LongCounter initOrGetCounterMetric(final ThreadContext context,
         return retrievedMetric.toJava(LongCounter.class);
     }
 
+    private Optional<NumberGauge> initOrGetNumberGaugeMetric(final ThreadContext context,
+                                                             final RubySymbol[] subPipelineNamespacePath,
+                                                             final RubySymbol metricName) {
+        final IRubyObject collector = this.metric.collector(context);
+        final IRubyObject fullNamespace = pipelineNamespacedPath(subPipelineNamespacePath);
+        final IRubyObject retrievedMetric = collector.callMethod(context, "get", new IRubyObject[]{fullNamespace, metricName, context.runtime.newSymbol("gauge")});
+
+        LazyDelegatingGauge delegatingGauge = retrievedMetric.toJava(LazyDelegatingGauge.class);
+        if (Objects.isNull(delegatingGauge.getType()) ||
+                MetricType.GAUGE_NUMBER.asString().equals(delegatingGauge.getType().asString()) == false) {
+            return Optional.empty();
+        }
+
+        NumberGauge castedNumberGauge = (NumberGauge) delegatingGauge.getMetric().get();
+        return Optional.of(castedNumberGauge);
```

Comment:
We can use the existing immutable `Optional` returned from `LazyDelegatingGauge#getMetric`:

```suggestion
        //noinspection unchecked
        return delegatingGauge.getMetric();
```

Reply:
`delegatingGauge.getMetric()` may not be `NumberGauge`. The casting would be be safeguard to intentionally break metric initialization here. Method itself also clarifies by name that it promises to return numeric gauge type.

Reply:
Updated!